qiskit_machine_learning/neural_networks/circuit_qnn.py

```diff
@@ -68,51 +68,61 @@ def __init__(self, circuit: QuantumCircuit,
         # TODO: need to be able to handle partial measurements! (partial trace...)
         self._circuit = circuit.copy()
-        if quantum_instance.is_statevector:
-            if len(self._circuit.clbits) > 0:
+        if self._quantum_instance.is_statevector:
+            if self._circuit.num_clbits > 0:
                 self._circuit.remove_final_measurements()
-        elif len(self._circuit.clbits) == 0:
+        elif self._circuit.num_clbits == 0:
             self._circuit.measure_all()
 
-        self._input_params = list(input_params or [])
-        self._weight_params = list(weight_params or [])
-        self._interpret = interpret if interpret else lambda x: x
-        sparse_ = False if sampling else sparse
         output_shape_ = self._compute_output_shape(interpret, output_shape, sampling)
 
+        # init super class
+        super().__init__(len(self._input_params), len(self._weight_params), sparse_, sampling,
+                         output_shape_)
+
+        # prepare sampler
+        self._sampler = CircuitSampler(self._quantum_instance, param_qobj=False, caching='all')
+
+        self._original_circuit = circuit
         # use given gradient or default
-        self._gradient = gradient if gradient else Gradient()
+        self._gradient = gradient or Gradient()
 
-        # construct probability gradient opflow object
-        self._grad_circuit: QuantumCircuit = None
+        # prepare probability gradient opflow object
+        self._construct_gradient_circuit()
+
+    def _construct_gradient_circuit(self):
+        self._gradient_circuit: QuantumCircuit = None
         try:
-            grad_circuit = circuit.copy()
+            grad_circuit = self._original_circuit.copy()
             grad_circuit.remove_final_measurements()  # ideally this would not be necessary
-            params = list(input_params) + list(weight_params)
-            self._grad_circuit = self._gradient.convert(StateFn(grad_circuit), params)
+            if self._input_gradients:
+                params = self._input_params + self._weight_params
+            else:
+                params = self._weight_params
+            self._gradient_circuit = self._gradient.convert(StateFn(grad_circuit), params)
         except (ValueError, TypeError, OpflowError, QiskitError):
             logger.warning('Cannot compute gradient operator! Continuing without gradients!')
 
-        super().__init__(len(self._input_params), len(self._weight_params), sparse_, sampling,
-                         output_shape_)
-
     def _compute_output_shape(self, interpret, output_shape, sampling) -> Tuple[int, ...]:
         """Validate and compute the output shape."""
 
         # this definition is required by mypy
         output_shape_: Tuple[int, ...] = (-1,)
         if sampling:
-            num_samples = self.quantum_instance.run_config.shots
+            num_samples = self._quantum_instance.run_config.shots
```

Comment:
Is there a rationale to change this from accessing the quantity via the public property to using the private variable that is currently backing it? Same for `self.circuit` change below

Reply:
Only one: if possible, reduce a number of abstractions. In case of `self._*` I know this is a plain old private class member. In case of a property you never know what is built in it, e.g. properties of `BlueprintCircuit`, this is an example of how properties should NOT be designed in my opinion.

Reply:
Arguably the getter, like it is here, should not have any side-effects, unlike a setter where sometimes other things are invalidated or changed via the property setter that will not happen if you use the variable that is backing it. I have seen things fail because of this when the property should ideally have been used consistently. If a property exists my choice is normally to use it unless we have to avoid any side-effect.

Reply:
A lesson learnt from `BlueprintCircuit` is that even in a getter can be some logic with side effects, e.g. `parameters`. Properties are for public access, internally, in the code, there should not be a difference and plain member access is just simpler.

Reply:
With a mix you just have to be more careful evidently - and people changing the code have to take care. As to simpler, arguably I think its simpler if they all go through the getter/setter, rather than a mix. As I mentioned I have seen things go wrong because of direct access when side-effect was desired. Seems you have too but in an opposite way. Either way with the code doing it 2 different ways one just has to take more care.

Reply:
Oh... I see your point now. Why I did not notice this earlier. There should be a consistent way of accessing class members. I will review the code! Thanks for spotting it!

Reply:
Now class member access is consistent.